ssh-server/main.go

```diff
@@ -113,38 +118,106 @@ func serviceChannel(newChannel ssh.NewChannel) error {
+		case <-done:
+			return nil
+		case req, ok := <-requests:
+			if !ok {
+				return nil
+			}
+			if req.WantReply {
+				req.Reply(allow[req.Type], nil)
+			}
+			switch req.Type {
+			case "pty-req":
+				r := bytes.NewReader(req.Payload)
+				term, err := parseString(r)
+				if err != nil {
+					return fmt.Errorf("parsing TERM environment variable value: %v", err)
+				}
+
+				size, err := windowSize(r)
+				if err != nil {
+					return fmt.Errorf("parsing window size: %v", err)
+				}
+				resize <- size
+
+				go func() {
+					runPTY(channel, term, resize)
+					close(done)
+				}()
+			case "window-change":
+				size, err := windowSize(bytes.NewReader(req.Payload))
+				if err != nil {
+					log.Printf("Failed to parse window size: %v", err)
+					continue
+				}
+				resize <- size
+			}
+		}
+	}
+}
+
+func runPTY(channel ssh.Channel, term string, resize chan *pty.Winsize) {
 	// Start the command with a pseudo-terminal.
-	shell, err := pty.Start(exec.Command("bash"))
+	cmd := exec.Command("bash")
+	cmd.Env = append(cmd.Env, fmt.Sprintf("TERM=%s", term))
+	shell, err := pty.Start(cmd)
 	if err != nil {
-		return fmt.Errorf("starting pty: %v", err)
+		log.Printf("Failed to start pty: %v", err)
+		return
 	}
 	defer shell.Close()
 
-	go func(in <-chan *ssh.Request) {
-		for req := range in {
-			// Tell the client we accept pty and shell commands
-			req.Reply(req.Type == "pty-req" || req.Type == "shell", nil)
-
-			if req.Type == "pty-req" {
-				skip := binary.BigEndian.Uint32(req.Payload)
-				resize(shell, req.Payload[4+skip:])
-			}
-			if req.Type == "window-change" {
-				resize(shell, req.Payload)
-			}
+	go func() {
+		for size := range resize {
+			pty.Setsize(shell, size)
 		}
-	}(requests)
+	}()
 
 	go io.Copy(shell, channel)
 	io.Copy(channel, shell)
-	return nil
+
+	status := struct{ Status uint32 }{}
+	if _, err := channel.SendRequest("exit-status", false, ssh.Marshal(&status)); err != nil {
+		log.Printf("Failed to send exit status: %v", err)
+	}
+}
+
+func windowSize(r io.Reader) (*pty.Winsize, error) {
+	width, err := parseUint32(r)
+	if err != nil {
+		return nil, fmt.Errorf("parsing window width: %v", err)
+	}
+	height, err := parseUint32(r)
+	if err != nil {
+		return nil, fmt.Errorf("parsing window height: %v", err)
+	}
+	return &pty.Winsize{Cols: uint16(width), Rows: uint16(height)}, nil
 }
 
-func resize(f *os.File, payload []byte) {
-	var size struct {
-		Width, Height uint32
+func parseUint32(r io.Reader) (uint32, error) {
+	var param uint32
+	if err := binary.Read(r, binary.BigEndian, param); err != nil {
+		return 0, err
 	}
-	if err := binary.Read(bytes.NewReader(payload), binary.BigEndian, &size); err == nil {
-		pty.Setsize(f, &pty.Winsize{Cols: uint16(size.Width), Rows: uint16(size.Height)})
+	return param, nil
+}
+
+func parseString(r io.Reader) (string, error) {
+	length, err := parseUint32(r)
+	if err != nil {
+		return "", fmt.Errorf("parsing length: %v", err)
+	}
+
+	term := make([]byte, length)
```

Comment:
str instead of term, since this is a generic function.

Reply:
done